File: test/dump_policy.py
# Dump every policy decision the python implementation makes, for parity checking.
import importlib.util, json, os, sys
here = os.path.dirname(os.path.abspath(__file__))
spec = importlib.util.spec_from_loader("cli", None)
cli = importlib.util.module_from_spec(spec)
exec(open(os.path.join(here, "..", "bin", "omarchy-spaces")).read().replace(
    'if __name__ == "__main__":\n    sys.exit(main(sys.argv[1:]))', ''), cli.__dict__)

cfg = json.load(open(os.path.join(here, "..", "config", "spaces.example.json")))
apps = ["Signal", "Slack", "Ghost", "spotify", "SLACK.desktop", ""]
out = []
for sid in cli.space_ids(cfg):
    for m in range(1440):
        pol = cli.resolve_policy(cfg, sid, m)
        row = [sid, m, sorted(pol["allowFrom"]), pol["source"], bool(pol["allowUnassigned"])]
        for a in apps:
            owner = cli.space_for_app(cfg, a)
            allowed = (bool(pol["allowUnassigned"]) if owner is None
                       else owner in pol["allowFrom"])
            row.append([a, owner, allowed])
        out.append(row)
print(json.dumps(out, separators=(",", ":")))

File: test/dump_policy.js
// Dump every policy decision the JS implementation makes, for parity checking.
var fs = require("fs"), path = require("path")
var L = require("../SpacesLogic.js")
var cfg = JSON.parse(fs.readFileSync(path.join(__dirname, "..", "config", "spaces.example.json"), "utf8"))
var apps = ["Signal", "Slack", "Ghost", "spotify", "SLACK.desktop", ""]
var out = []
L.spaceIds(cfg).forEach(function (sid) {
  for (var m = 0; m < 1440; m++) {
    var pol = L.resolvePolicy(cfg, sid, m)
    var row = [sid, m, pol.allowFrom.slice().sort(), pol.source, !!pol.allowUnassigned]
    apps.forEach(function (a) {
      var owner = L.spaceForApp(cfg, a)
      var allowed = owner === null ? !!pol.allowUnassigned : pol.allowFrom.indexOf(owner) !== -1
      row.push([a, owner, allowed])
    })
    out.push(row)
  }
})
console.log(JSON.stringify(out))
